kolibri/core/logger/api.py

```diff
@@ -1,192 +1,774 @@
+
+
+class AttemptSerializer(serializers.Serializer):
+    id = HexStringUUIDField(required=False)
+    item = serializers.CharField(required=False)
+    correct = serializers.FloatField(min_value=0, max_value=1)
+    complete = serializers.BooleanField(required=False, default=False)
+    time_spent = serializers.FloatField(min_value=0)
+
+    answer = serializers.DictField(required=False)
+    simple_answer = serializers.CharField(required=False, allow_blank=True)
+    error = serializers.BooleanField(required=False, default=False)
+    hinted = serializers.BooleanField(required=False, default=False)
+    # Whether to replace the current answer with the new answer
+    # this is a no-op if the attempt is being created.
+    replace = serializers.BooleanField(required=False, default=False)
+
+    # An additional field that can be set to handle coach assigned quizzes
+    # that are themselves not proper content nodes but refer to content under
+    # the hood.
+    content_id = HexStringUUIDField(required=False)
+
+    def validate(self, data):
+        if not data["error"] and "answer" not in data:
+            raise ValidationError("Must provide an answer if not an error")
+        if "id" not in data and "item" not in data:
+            raise ValidationError("Must provide an id or item")
+        return data
+
+
+class UpdateSessionSerializer(serializers.Serializer):
+    progress_delta = serializers.FloatField(min_value=0, max_value=1.0, required=False)
+    progress = serializers.FloatField(min_value=0, max_value=1.0, required=False)
+    time_spent_delta = serializers.FloatField(min_value=0, required=False)
+    extra_fields = serializers.DictField(required=False)
+    attempt = AttemptSerializer(required=False)
+
+    def validate(self, data):
+        if "progress_delta" in data and "progress" in data:
+            raise ValidationError(
+                "must not pass progress_delta and progress in the same request"
+            )
+        return data
+
+
+attemptlog_fields = [
+    "id",
+    "correct",
+    "complete",
+    "hinted",
+    "error",
+    "item",
+    "answer",
+    "time_spent",
+]
+
+
+def serialize_quiz_attempt_log(log):
+    try:
+        item_content_id, item = log["item"].split(QUIZ_ITEM_DELIMETER)
+        log["item"] = item
+    except ValueError:
+        item_content_id = ""
+    log["content_id"] = item_content_id
+    return log
+
+
+class ProgressTrackingViewSet(viewsets.GenericViewSet):
+    def _validate_quiz(self, user, quiz_id):
+        if user.is_anonymous():
+            raise PermissionDenied("Cannot access a quiz if not logged in")
+        if not Exam.objects.filter(
+            active=True,
+            assignments__collection_id__in=user.memberships.all().values(
+                "collection_id"
+            ),
+            id=quiz_id,
+        ).exists():
+            raise PermissionDenied("User does not have access to this quiz_id")
+
+    def _validate_context(self, user, validated_data):
```

Comment:
since this returns values, it does more than just validation

```suggestion
    def _extract_context(self, user, validated_data):
```

Reply:
True.